### DISCUSSION TITLE
[Help]: React testing library with embla-carousel

### DISCUSSION THREAD

**Discussion by SzymonGalazka** [davidjerleke/embla-carousel#775] [Help] [answered]
### Summary

Hi everyone.
I'm trying to test a carousel built on top of `embla-carousel-8.0.0` using react testing library.
I encountered a problem that was mentioned [here](https://github.com/davidjerleke/embla-carousel/issues/516), and I was able to render `embla-carousel` by mocking the following items:

```ts
// mock matchMedia
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// mock IntersectionObserver
class IntersectionObserver {
  observe = jest.fn();
  disconnect = jest.fn();
  unobserve = jest.fn();
}

Object.defineProperty(window, 'IntersectionObserver', {
  writable: true,
  configurable: true,
  value: IntersectionObserver,
});

// mock ResizeObserver
class ResizeObserver {
  observe = jest.fn();
  unobserve = jest.fn();
  disconnect = jest.fn();
}
Object.defineProperty(window, 'ResizeObserver', {
  writable: true,
  configurable: true,
  value: ResizeObserver,
});
```

The carousel **does render**, but the internal calculations from the `useEmblaCarousel()` are all wrong and I cannot properly test it from user perspective (which is the testing-library way).

For example, after I render a carousel, I want to:
1. test if I can interact with the 'next' button when there are more slides than in view
2. test if I cannot interact with the 'next' button when I reached the end of the slides

This is not possible as `nextDisabled, prevDisabled` props are returning wrong values in test environment (works as expected in the app).

I wonder if there is a working example of how the mocks should be set up to enable such tests.

### If applicable, which variants of Embla Carousel are relevant to this question?

- [X] embla-carousel **(Core)**
- [X] embla-carousel-react
- [ ] embla-carousel-vue
- [ ] embla-carousel-svelte
- [ ] embla-carousel-autoplay
- [ ] embla-carousel-auto-scroll
- [ ] embla-carousel-solid
- [ ] embla-carousel-auto-height
- [ ] embla-carousel-class-names
- [ ] embla-carousel-docs **(Documentation)**
- [ ] embla-carousel-docs **(Generator)**

**Comment by davidjerleke** [comment, marked as answer]
> The carousel **does render**, but the internal calculations from the `useEmblaCarousel()` are all wrong and I cannot properly test it from user perspective (which is the testing-library way).
>
> For example, after I render a carousel, I want to:
> 1. test if I can interact with the 'next' button when there are more slides than in view
> 2. test if I cannot interact with the 'next' button when I reached the end of the slides
>
> This is not possible as `nextDisabled, prevDisabled` props are returning wrong values in test environment (works as expected in the app).
>
> I wonder if there is a working example of how the mocks should be set up to enable such tests.



Hi @SzymonGalazka,

Thanks for your question. If I'm not mistaken, your test environment isn't a real browser (unless you're using Cypress or similar). And the problem with non browser environments is that elements won't have any dimensions. This is expected behaviour.

The `embla-carousel` library solves these problems by **mocking element dimensions**. All test files [import](https://github.com/davidjerleke/embla-carousel/blob/master/packages/embla-carousel/src/__tests__/constructor.test.ts#L2-L5) a function called [`mockTestElements`](https://github.com/davidjerleke/embla-carousel/blob/master/packages/embla-carousel/src/__tests__/mocks/testElements.mock.ts#L51-L64) and relevant [fixtures](https://github.com/davidjerleke/embla-carousel/blob/master/packages/embla-carousel/src/__tests__/fixtures/constructor.fixture.ts#L10-L53).

Here's the process outlined:

1. I create a carousel with relevant options for the test and open it in a real browser.
2. I read the `offsetWidth`, `offsetHeight`, `offsetTop`, `offsetLeft` values for the **container** and **slide elements** respectively (I use a script to automatically do this but it can be done manually too of course).
3. I create a **fixture** with the **offset values** like [this](https://github.com/davidjerleke/embla-carousel/blob/master/packages/embla-carousel/src/__tests__/fixtures/constructor.fixture.ts#L10).
4. I pass the **fixtures** to the [`mockTestElements`](https://github.com/davidjerleke/embla-carousel/blob/master/packages/embla-carousel/src/__tests__/mocks/testElements.mock.ts#L51-L64) function which simply assigns all elements relevant dimensions in my test environment just like they have in the real browser.

I hope you find this information useful.

Best,
David

**Comment by SzymonGalazka** [comment]
Thanks @davidjerleke that makes perfect sense from a library perspective.

However, this approach directly passes mocks to the carousel hook, and the mocks are carousel-specific. This way you're testing it internally.
Ideally, from a testing-library perspective, mock should be just a set of webAPI mocks relevant to this library applied on the environment (e.g. jest). Then, the carousel should be able to render in an environment prepared this way. 
The thing is I'm not sure how to setup the environment to meet library's needs - and asking to see if someone has done that.

Feel free to close it if you think it's not relevant to the library per se.

Thanks for your great work anyway.

**Reply by DanielSpindler** [reply]
Hey @SzymonGalazka ,

ive encounterd the same problem in Playwright and "fixxed" it with using a production build (which is better anyways as i found out) instead of the dev enviroment
